system/src/system_info.cpp

```diff
@@ -0,0 +1,407 @@
+		return json.getState();
+	}
+
+	inline bool closeDocument() {
+		json.endObject();
+		return json.getState();
+	}
+
+	bool formatSourceError(const diag_source* src, int error) {
+		json.name(src->name);
+		json.beginObject();
+		json.name("err").value(error);
+		json.endObject();
+		return json.getState();
+	}
+
+	inline bool isSourceOk(const diag_source* src) {
+	    return (src->name);
+	}
+
+	inline bool formatSourceInt(const diag_source* src, AbstractIntegerDiagnosticData::IntType val) {
+		json.name(src->name).value(val);
+		return json.getState();
+	}
+
+	inline bool formatSourceUnsignedInt(const diag_source* src, AbstractUnsignedIntegerDiagnosticData::IntType val) {
+		json.name(src->name).value(val);
+		return json.getState();
+	}
+};
+
+
+class BinaryDiagnosticsFormatter : public AbstractDiagnosticsFormatter<BinaryDiagnosticsFormatter> {
+
+	AppendData& data;
+
+	// FIXME: single size for all the diagnostics is just plain wrong
+	using value = AbstractIntegerDiagnosticData::IntType;
+	using id = typeof(diag_source::id);
+
+public:
+	BinaryDiagnosticsFormatter(AppendData& appender_) : data(appender_) {}
+
+
+	inline bool openDocument() {
+		return data.write(uint16_t(sizeof(id))) && data.write(uint16_t(sizeof(value)));
+	}
+
+	inline bool closeDocument() {
+		return true;
+	}
+
+	/**
+	 *
+	 */
+	bool formatSourceError(const diag_source* src, int error) {
+		static_assert(sizeof(src->id)==2, "expected diagnostic id to be 16-bits");
+		return data.write(decltype(src->id)(src->id | 1<<15)) && data.write(int32_t(error));
+	}
+
+	inline bool isSourceOk(const diag_source* src) {
+	    return true;
+	}
+
+	inline bool formatSourceInt(const diag_source* src, AbstractIntegerDiagnosticData::IntType val) {
+		return data.write(src->id) && data.write(val);
+	}
+
+	inline bool formatSourceUnsignedInt(const diag_source* src, AbstractUnsignedIntegerDiagnosticData::IntType val) {
+		return data.write(src->id) && data.write(val);
+	}
+
+};
+
+} // anonymous
+
+bool module_info_to_json(AppendJson& json, const hal_module_t* module, uint32_t flags)
+{
+    const module_info_t* info = &module->info;
+
+    bool output_uuid = (module_function(info) == MODULE_FUNCTION_USER_PART);
+    json.beginObject();
+    json.name("s").value(module->bounds.maximum_size);
+    json.name("l").value(module_store_string(module->bounds.store));
+    json.name("vc").value(module->validity_checked);
+    json.name("vv").value(module->validity_result);
+    if (output_uuid) {
+        char buf[sizeof(module_info_suffix_t::sha) * 2] = {};
+        json.name("u").value(bytes2hexbuf(module->suffix.sha, sizeof(module->suffix.sha), buf), sizeof(buf));
+    }
+    // FIXME?
+    if (info) {
+        json.name("f").value(module_function_string(module_function(info)));
+        char tmp[4] = {};
+        snprintf(tmp, sizeof(tmp), "%u", (unsigned)module_index(info));
+        // FIXME: is this really supposed to be a string?
+        // json.name("n").value(module_index(info));
```

Comment:
@sergeuz Can we make this an integer instead of a string? Does DS care about this? CLI seems to take it okay. As of right now the describe with factory module on an Electron takes 800 bytes which is exactly MAX_CONTENT_LEN on 2.x, with this turned into an int we end up with 776 instead.

Reply:
(some notes from the discussion)
Looks like we'd better not, so instead when backporting to 2.x, we'll increase `MBEDTLS_SSL_MAX_CONTENT_LEN` and `PROTOCOL_BUFFER_SIZE` to 900 on Electron specifically.